File: app/src/main/java/com/siea/prayerdisplay/domain/PrayerEngine.kt
package com.siea.prayerdisplay.domain

import com.siea.prayerdisplay.data.ContentEntity
import com.siea.prayerdisplay.data.PrayerDayEntity
import java.time.Duration
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZonedDateTime

enum class PrayerName(val displayName: String) {
    FAJR("FAJR"), DHUHR("DHUHR"), ASR("ASR"), MAGHRIB("MAGHRIB"), ISHA("ISHA")
}

data class PrayerOccurrence(
    val name: PrayerName,
    val date: LocalDate,
    val adhan: ZonedDateTime,
    val iqamah: ZonedDateTime
)

data class PrayerRow(
    val name: PrayerName,
    val adhan: String,
    val iqamah: String,
    val isTomorrow: Boolean,
    val isNext: Boolean
)

sealed interface DisplayState {
    data object Loading : DisplayState
    data class Main(val rows: List<PrayerRow>) : DisplayState
    data class Adhan(val prayer: PrayerName, val time: String) : DisplayState
    data class IqamahCountdown(val prayer: PrayerName, val secondsRemaining: Long) : DisplayState
    data class DailyContent(val item: ContentEntity) : DisplayState
}

object PrayerEngine {
    private const val IQAMAH_COUNTDOWN_SECONDS = 120L
    private const val ADHAN_SCREEN_SECONDS = 60L
    private const val CONTENT_QUIET_AFTER_IQAMAH_MINUTES = 15L

    fun evaluate(
        now: ZonedDateTime,
        days: List<PrayerDayEntity>,
        content: List<ContentEntity>
    ): DisplayState {
        if (days.isEmpty()) return DisplayState.Loading
        val occurrences = days.flatMap { it.occurrences(now) }

        occurrences.firstOrNull {
            !now.isBefore(it.iqamah.minusSeconds(IQAMAH_COUNTDOWN_SECONDS)) && now.isBefore(it.iqamah)
        }?.let {
            return DisplayState.IqamahCountdown(
                prayer = it.name,
                secondsRemaining = Duration.between(now, it.iqamah).seconds.coerceAtLeast(0)
            )
        }

        occurrences.firstOrNull {
            !now.isBefore(it.adhan) && now.isBefore(it.adhan.plusSeconds(ADHAN_SCREEN_SECONDS))
        }?.let { return DisplayState.Adhan(it.name, it.adhan.toLocalTime().toClockText()) }

        val inQuietWindow = occurrences.any {
            !now.isBefore(it.iqamah.minusSeconds(IQAMAH_COUNTDOWN_SECONDS)) &&
                now.isBefore(it.iqamah.plusMinutes(CONTENT_QUIET_AFTER_IQAMAH_MINUTES))
        }
        val contentMoment = now.minute % 5 == 0 && now.second < 30
        if (!inQuietWindow && contentMoment) {
            val todayContent = content.filter { it.date == now.toLocalDate().toString() || it.date == "*" }
            if (todayContent.isNotEmpty()) {
                val index = (now.toLocalTime().toSecondOfDay() / 300) % todayContent.size
                return DisplayState.DailyContent(todayContent[index])
            }
        }

        return DisplayState.Main(buildRows(now, occurrences))
    }

    private fun buildRows(now: ZonedDateTime, occurrences: List<PrayerOccurrence>): List<PrayerRow> {
        val nextIqamah = occurrences.filter { now.isBefore(it.iqamah) }.minByOrNull { it.iqamah }
        return PrayerName.entries.map { name ->
            val next = occurrences
                .filter { it.name == name && now.isBefore(it.iqamah) }
                .minByOrNull { it.iqamah }
            PrayerRow(
                name = name,
                adhan = next?.adhan?.toLocalTime()?.toClockText() ?: "--:--",
                iqamah = next?.iqamah?.toLocalTime()?.toClockText() ?: "--:--",
                isTomorrow = next?.date?.isAfter(now.toLocalDate()) == true,
                isNext = next != null && next == nextIqamah
            )
        }
    }

    private fun PrayerDayEntity.occurrences(now: ZonedDateTime): List<PrayerOccurrence> {
        val day = LocalDate.parse(date)
        fun occurrence(name: PrayerName, adhanText: String, iqamahText: String) = PrayerOccurrence(
            name = name,
            date = day,
            adhan = ZonedDateTime.of(day, LocalTime.parse(adhanText), now.zone),
            iqamah = ZonedDateTime.of(day, LocalTime.parse(iqamahText), now.zone)
        )
        return listOf(
            occurrence(PrayerName.FAJR, fajrAdhan, fajrIqamah),
            occurrence(PrayerName.DHUHR, dhuhrAdhan, dhuhrIqamah),
            occurrence(PrayerName.ASR, asrAdhan, asrIqamah),
            occurrence(PrayerName.MAGHRIB, maghribAdhan, maghribIqamah),
            occurrence(PrayerName.ISHA, ishaAdhan, ishaIqamah)
        )
    }

    private fun LocalTime.toClockText(): String = "%02d:%02d".format(hour, minute)
}

File: web/prayer-engine.js
(function (root, factory) {
  const api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;
  root.PrayerEngine = api;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  const PRAYERS = ["fajr", "dhuhr", "asr", "maghrib", "isha"];
  const LABELS = { fajr: "FAJR", dhuhr: "DHUHR", asr: "ASR", maghrib: "MAGHRIB", isha: "ISHA" };
  const ADHAN_SCREEN_SECONDS = 60;
  const IQAMAH_COUNTDOWN_SECONDS = 120;
  const QUIET_AFTER_IQAMAH_SECONDS = 15 * 60;

  function localStamp(date, time = "00:00", seconds = 0) {
    const [year, month, day] = date.split("-").map(Number);
    const [hour, minute] = time.split(":").map(Number);
    return Date.UTC(year, month - 1, day, hour, minute, seconds);
  }

  function addDays(date, amount) {
    const value = new Date(localStamp(date));
    value.setUTCDate(value.getUTCDate() + amount);
    return value.toISOString().slice(0, 10);
  }

  function nowStamp(parts) {
    return localStamp(parts.date, `${parts.hour}:${parts.minute}`, Number(parts.second || 0));
  }

  function occurrences(days) {
    return days.flatMap((day) => PRAYERS.map((name) => ({
      name,
      label: LABELS[name],
      date: day.date,
      adhanText: day[name].adhan,
      iqamahText: day[name].iqamah,
      adhan: localStamp(day.date, day[name].adhan),
      iqamah: localStamp(day.date, day[name].iqamah),
    }))).sort((a, b) => a.iqamah - b.iqamah);
  }

  function eligibleContent(content, date) {
    return content.filter((item) => item.date === "*" || item.date === date);
  }

  function buildRows(current, all) {
    const nextIqamah = all.filter((item) => current < item.iqamah).sort((a, b) => a.iqamah - b.iqamah)[0];
    return PRAYERS.map((name) => {
      const next = all.filter((item) => item.name === name && current < item.iqamah).sort((a, b) => a.iqamah - b.iqamah)[0];
      return {
        name,
        label: LABELS[name],
        adhan: next?.adhanText || "--:--",
        iqamah: next?.iqamahText || "--:--",
        date: next?.date || null,
        isTomorrow: Boolean(next && next.date > new Date(current).toISOString().slice(0, 10)),
        isNext: Boolean(next && nextIqamah && next.name === nextIqamah.name && next.date === nextIqamah.date),
      };
    });
  }

  function evaluate(parts, payload) {
    if (!payload?.prayer_days?.length) return { type: "loading" };
    const current = nowStamp(parts);
    const all = occurrences(payload.prayer_days);

    // Adhan takes precedence where Maghrib Iqamah is only two minutes later,
    // ensuring the one-minute Adhan notice is still visible.
    const adhan = all.find((item) => current >= item.adhan && current < item.adhan + ADHAN_SCREEN_SECONDS * 1000);
    if (adhan) return { type: "adhan", prayer: adhan.label, time: adhan.adhanText };

    const countdown = all.find((item) => current >= item.iqamah - IQAMAH_COUNTDOWN_SECONDS * 1000 && current < item.iqamah);
    if (countdown) {
      return {
        type: "iqamah",
        prayer: countdown.label,
        secondsRemaining: Math.max(0, Math.ceil((countdown.iqamah - current) / 1000)),
      };
    }

    const quiet = all.some((item) => current >= item.iqamah - IQAMAH_COUNTDOWN_SECONDS * 1000 && current < item.iqamah + QUIET_AFTER_IQAMAH_SECONDS * 1000);
    const minute = Number(parts.minute);
    const second = Number(parts.second || 0);
    const contentMoment = minute % 5 === 0 && second < 30;
    if (!quiet && contentMoment) {
      const items = eligibleContent(payload.content || [], parts.date);
      if (items.length) {
        const secondsOfDay = Number(parts.hour) * 3600 + minute * 60 + second;
        return { type: "content", item: items[Math.floor(secondsOfDay / 300) % items.length] };
      }
    }

    return { type: "main", rows: buildRows(current, all) };
  }

  return { PRAYERS, LABELS, addDays, evaluate, localStamp, occurrences };
});
